### DISCUSSION TITLE
Variable importance figure printing weird

### DISCUSSION THREAD

**Discussion by morbenk** [kogalur/randomForestSRC#378] [Q&A] [answered]
I am looking to use Random forest for an imbalanced class problem. the following code was taken directly off the example from the "Random Forests Quantile Classifier" vignette. When I print the results I get a confusion matrix class.error rate of 0, this is my first question: is this a possible answer?  

RF<-rfsrc(Presence ~ ., LKC, rfq =  TRUE, ntree = 3000, perf.type = "g.mean", importance = TRUE)
print(RF)

> print(RF)
                         Sample size: 71
           Frequency of class labels: 57, 14
                     Number of trees: 3000
           Forest terminal node size: 1
       Average no. of terminal nodes: 8.319
No. of variables tried at each split: 6
              Total no. of variables: 30
       Resampling used to grow trees: swor
    Resample size used to grow trees: 45
                            Analysis: RFQ
                              Family: class
                      Splitting rule: gini *random*
       Number of random split points: 10
                    Imbalanced ratio: 4.0714
                   (OOB) Brier score: 0.12174667
        (OOB) Normalized Brier score: 0.48698669
                           (OOB) AUC: 0.89348371
                        (OOB) PR-AUC: 0.48743056
                        (OOB) G-mean: 0.88852332
   (OOB) Requested performance error: 0.11147668

Confusion matrix:

          predicted
  observed  0  1 class.error
         0 45 12      0.2105
         1  0 14      0.0000

      (OOB) Misclassification rate: 0.1690141

Then when I plot the graph I come up with the following figure.... I must be setting my data up wrong and I could not find the "breast" dataset that was used in the vignette. Would anyone be able to give me some insight. 

plot(RF, plots.one.page = FALSE)

Thanks from a barely surviving grad student who just wants to be finished. 

K 

![RF](https://user-images.githubusercontent.com/132721526/236589500-67cf5e09-e4ad-46a0-9a12-d263bf3ed88b.png)

**Comment by morbenk** [comment]
Also, new to github and how to use its discussion features sorry for the clunky post.

**Comment by ishwaran** [comment, marked as answer]
Answer to your first question: yes in class imbalanced settings the misclassification error rate can be very small (in your case 0.111) even sometimes nearly zero. This is misleading. All standard classifiers use Bayes rule which predicts the class label as 1 if the probability is greater than 0.5.  But if you have imbalanced data this is very unlikely to happen since the probability of being class label 1 is very low (because most of the data is class 0 -- that's why it's imbalanced!!) and therefore the classifier will assign most (if not all data) to class label 0.  But that will look deceptively good if you measure performance by misclassification error. 

This is why you should not use the confusion matrix as a measure of performance.  Same goes for AUC, Brier.  Best metrics are PR-AUC and gmean.

Secondly, I recommend directly working with the function `imbalanced` for class imbalanced data which implements RFQ.  See the [help file](https://www.randomforestsrc.org/reference/imbalanced.rfsrc.html)

Finally,  to access the breast cancer data (also see help file above):

```
data(breast, package = "randomForestSRC.stealth")
breast <- na.omit(breast)
```

**Reply by morbenk** [reply]
Thank you very much. This helps a lot. 

One question that came up as well, after reading the VIMP vignette, it looks like the best variable importance procedure would be the subsample procedure that uses the delete-d jackknife estimator to calculate confidence intervals, is that correct? Then I can decide using: 

extract.subsample(variables)$var.jk.sel.Z

which variables have the desired p-value? 

Thanks, 
K

**Comment by ishwaran** [comment]
Your confusing things.

The subsampling procedures are for getting standard errors + confidence intervals for variable importance.

For imbalanced data the best variable importance is gmean.

To get gmean variable importance and subsampled standard errors + confidence intervals:


```
o <- imbalanced(my.formula, mydata, importance = TRUE, block.size = 10)
smp.o <- subsample(o, B = 100)
plot.subsample(smp.o)
```

**Reply by morbenk** [reply]
Got it. Thank you for all your help.
